File: src/lib/scraper/states/fl-dep.ts
/**
 * Florida — Department of Environmental Protection, Storage Tank facilities.
 *
 * FL DEP publishes the full statewide storage-tank facility roster through
 * its Open Data portal as a single CSV (~6 MB):
 *   https://geodata.dep.state.fl.us/datasets/storage-tank-contamination-monitoring-facilities.csv
 *
 * Columns of interest (snake-case in the export):
 *   facility_id, facility_name, owner_name, owner_phone, owner_email,
 *   facility_address, facility_city, facility_county, facility_zip,
 *   tank_count, regulatory_status
 *
 * The DepStorage HTML search at prodenv.dep.state.fl.us/DepStorage/ is also
 * available for ad-hoc lookups, but the bulk CSV is canonical and comes from
 * the same backing system, so we use it for the target-list build.
 */

import { fetchText, parseCsv, type FetchOptions } from '../http';
import type { RawFacilityRecord, ScraperRunOptions, StateScraper } from '../types';

const BULK_URL =
  'https://geodata.dep.state.fl.us/datasets/storage-tank-contamination-monitoring-facilities.csv';

function mapRow(row: Record<string, string>): RawFacilityRecord | null {
  const facilityName = row.facility_name || row.FACILITY_NAME;
  const facilityId = row.facility_id || row.FACILITY_ID;
  if (!facilityName || !facilityId) return null;

  const tanks = parseInt(row.tank_count || row.TANK_COUNT || '', 10);

  return {
    sourceFacilityId: facilityId,
    state: 'FL',
    facilityName,
    ownerName: row.owner_name || row.OWNER_NAME || undefined,
    address: row.facility_address || row.FACILITY_ADDRESS || undefined,
    city: row.facility_city || row.FACILITY_CITY || undefined,
    county: row.facility_county || row.FACILITY_COUNTY || undefined,
    zip: row.facility_zip || row.FACILITY_ZIP || undefined,
    phone: row.owner_phone || row.OWNER_PHONE || undefined,
    email: row.owner_email || row.OWNER_EMAIL || undefined,
    tankCount: Number.isFinite(tanks) ? tanks : undefined,
    facilityStatus: row.regulatory_status || row.REGULATORY_STATUS || undefined,
    sourceUrl: BULK_URL,
    raw: row,
  };
}

export const flDepScraper: StateScraper = {
  state: 'FL',
  agency: 'Florida Department of Environmental Protection',
  source: 'DEP Storage Tank facility bulk CSV',

  async fetchFacilities(options: ScraperRunOptions = {}): Promise<RawFacilityRecord[]> {
    const fetchOpts: FetchOptions = {
      fetcher: options.fetcher,
      timeoutMs: options.timeoutMs,
    };

    const text = await fetchText(BULK_URL, fetchOpts);
    const rows = parseCsv(text);

    const out: RawFacilityRecord[] = [];
    for (const row of rows) {
      const rec = mapRow(row);
      if (!rec) continue;
      if (
        options.activeOnly &&
        rec.facilityStatus &&
        !/open|active|regulated/i.test(rec.facilityStatus)
      ) {
        continue;
      }
      out.push(rec);
      if (options.limit && out.length >= options.limit) break;
    }

    return out;
  },
};

File: scrapers/sources/fl_dep.py
"""Florida — Department of Environmental Protection, Storage Tank facilities.

FL DEP publishes the full statewide storage-tank facility roster as a
single CSV on its open-data portal.

Endpoint: https://geodata.dep.state.fl.us/datasets/storage-tank-contamination-monitoring-facilities.csv
"""

from __future__ import annotations

import io

import pandas as pd
import requests

from ..common import canonicalize_columns

STATE = "FL"
URL = (
    "https://geodata.dep.state.fl.us/datasets/"
    "storage-tank-contamination-monitoring-facilities.csv"
)


def _address(row: pd.Series) -> str:
    parts = [
        row.get("facility_address"),
        row.get("facility_city"),
        STATE,
        row.get("facility_zip"),
    ]
    return ", ".join(p.strip() for p in parts if isinstance(p, str) and p.strip())


def fetch(session: requests.Session) -> pd.DataFrame:
    resp = session.get(URL)
    resp.raise_for_status()
    df = pd.read_csv(io.StringIO(resp.text), dtype=str).fillna("")

    # Accept either snake_case or UPPER_CASE headers; FL has flipped the
    # export style at least once in the last two years.
    lower = {c.lower(): c for c in df.columns}

    def col(name: str) -> pd.Series:
        actual = lower.get(name.lower())
        return df[actual] if actual else pd.Series([""] * len(df))

    out = pd.DataFrame(
        {
            "state": STATE,
            "facility_id": col("facility_id"),
            "facility_name": col("facility_name"),
            "owner_name": col("owner_name"),
            "owner_address": df.apply(_address, axis=1) if "facility_address" in lower else "",
            "owner_email_if_available": col("owner_email"),
            "last_inspection_date": col("last_inspection_date"),
            "violation_history": col("regulatory_status"),
        }
    )

    return canonicalize_columns(out, STATE, URL)
